jlm/rvsdg/control.cpp

```diff
@@ -41,13 +38,34 @@ ctltype::operator==(const jlm::rvsdg::type & other) const noexcept
   return type && type->nalternatives_ == nalternatives_;
 }
 
-std::unique_ptr<jlm::rvsdg::type>
+std::shared_ptr<const jlm::rvsdg::type>
 ctltype::copy() const
 {
-  return std::unique_ptr<jlm::rvsdg::type>(new ctltype(*this));
+  return std::make_shared<ctltype>(*this);
 }
 
-const ctltype ctl2(2);
+std::shared_ptr<const ctltype>
+ctltype::Create(std::size_t nalternatives)
+{
+  static const ctltype static_instances[4] = { // ctltype(0) is not valid, but put it in here so
+                                               // the static array indexing works correctly
+                                               ctltype(0),
+                                               ctltype(1),
+                                               ctltype(2),
+                                               ctltype(3)
+  };
+
+  if (nalternatives < 4)
+  {
+    return std::shared_ptr<const jlm::rvsdg::ctltype>(
+        std::shared_ptr<void>(),
+        &static_instances[nalternatives]);
+  }
+  else
+  {
+    return std::make_shared<ctltype>(nalternatives);
+  }
```

Comment:
Could we perform the check for `nalternaives == 0` here?

Reply:
done